### ISSUE TITLE
W0631: Using possibly undefined loop variable (undefined-loop-variable)

### ISSUE THREAD

**Issue by Clement-Lelievre** [PyCQA/pylint#8394]
### Bug description

_I've seen in the issues that similar claims have been made although not exactly this one unless I'm mistaken; apologies if this is redundant._

Sometimes you're sure the loop variable will exist, but Pylint still raises W0631 error **Using possibly undefined loop variable (undefined-loop-variable)**

MRE below:

```python
def func(numbers: list[int]) -> int | None:
    """docstring"""
    if not numbers: 
        return None
    for number in numbers:
        if number % 2:
            break
    return number  # [undefined-loop-variable]
```

The same is true if you `assert numbers` instead of `if not numbers` line 3.
FWIW, mypy doesn't complain with this piece of code

### Command used

```shell
pylint file.py
```

### Pylint output

```shell
************* Module file
file.py:15:11: W0631: Using possibly undefined loop variable 'number' (undefined-loop-variable)
```

### Expected behavior

I'd expect pylint **not** to raise the W0631 error message in those cases when we're certain that the returned variable will exist

### Pylint version

```shell
pylint 2.14.4
astroid 2.11.6
Python 3.10.0 (default, Dec  7 2021, 11:24:21) [GCC 9.3.0]
```

### OS / Environment

Ubuntu, ohmyZSH, WSL2

**Comment by nickdrozd** [comment]
Pylint is unable to infer the value of `numbers`, so it can't verify that the variable will be defined by the loop. Improved inference would fix this.
